src/utils/get-relative-date.ts

```diff
@@ -0,0 +1,42 @@
+function isSameDay(date1: Date, date2: Date): boolean {
+  return (
+    date1.getFullYear() === date2.getFullYear() &&
+    date1.getMonth() === date2.getMonth() &&
+    date1.getDate() === date2.getDate()
+  )
+}
+
+export function getRelativeDate(
+  startDateString: string,
+  endDateString: string
+): string {
+  const startDate = new Date(Date.parse(startDateString))
+  const endDate = new Date(Date.parse(endDateString))
+  const nowDate = new Date()
+
+  if (nowDate > endDate) return 'Proběhlo'
+  else if (nowDate > startDate) return 'Probíhá'
+  else {
+    let toReturn: string
+
+    if (isSameDay(startDate, nowDate)) toReturn = 'Dnes'
+    else {
+      const tomorrowDate = new Date(nowDate)
```

Comment:
Doporučuju nikdy neformátovat data a časy ručně, je to jedno velké klubko hadů :) Obecně platí, že s daty a časy je lepší dělat co nejvíc přes nabízená API, aby se člověk nespálil (off-by-one chyby, přestupné roky, změny času, časové zóny, etc). Silné povahy mohou nahlédnout [sem](https://gist.github.com/timvisee/fcda9bbdff88d45cc9061606b4b923ca) :) Takže tady asi doporučuju `Date.toLocaleString`, kde jsou nějaké vhodné presety.

A celkově je tohle výborný kandidát na pár unit testů? Přičemž (nejen) v rámci testů je dobré nečíst aktuální datum uvnitř funkce (`nowDate = new Date()`), ale předat ho jako parametr s výchozí hodnotou. Tak se pak dá v testech snadno poslat předem známá hodnota „now“.

Reply:
unit testy - souhlas

toLocaleString - zejména pokud chceme mít kontrolu nad tím jak je ten text dlouhý (viz ten bod že když je to celé na dva řádky, je to problém) - nemyslím si; jednak uživatelé jsou ze sta procent schopni pochopit tento datumový formát, a ostatní věci (přestupné roky, ...) tady nejsou problém - můžeš uvést konkrétní problém prosím?

Reply:
Nejde o to, že by ten formát byl nesrozumitelný nebo že by ten kód byl špatně. Ale je to špatný příklad, jako řezat na cirkulárce rukou blízko kotouče. Ty třeba víš, co děláš, ale odkouká to od tebe někdo druhý, kdo už to vědět nebude, a uřízne si prsty. Proto mně přijde lepší používat v těchto citlivých případech důsledně to „správné“ řešení, v tomhle případě formátovací API. Viz třeba:

```javascript
new Date().toLocaleString('cs-CZ', {month: "numeric", day: "numeric", "hour": "2-digit", minute: "2-digit"})
"26. 11. 13:28"
```

To je cajk, ne?

Reply:
Už to tak dělám, jen je podle mě škoda že tam ta funkce přidá mezeru - zbytečně to zabírá místo, osobně se mi to nelíbí a je to podle mě i nestandardní - díval jsem se na pár stránek (cnb.cz, mpo.cz) pro inspiraci a taky tam nemají mezeru. Vzhledem k tomu že to stejně stavím separátně pro datum a pro čas a navíc chci dávat podmíněně rok, tak by to vypadalo v kódu takto:

`...
toReturn = startDate.toLocaleString('cs-CZ', {month: 'numeric', day: 'numeric'});
        if (startDate.getFullYear() !== nowDate.getFullYear())
          toReturn += ' ' + startDate.getFullYear()
      }
    }

    toReturn += ', ' + startDate.toLocaleString('cs-CZ', {hour: '2-digit', minute: '2-digit'});
...`

což taky není žádná paráda, takže bych tam s dovolením asi přeci jen nechal to že si ten string postavím sám.

co myslíš? Hlavně ta mezera mě tam dráždí

Reply:
Koukal jsem na tu mezeru a [Pravidla připouští obě možnosti](https://prirucka.ujc.cas.cz/?id=810), přičemž jako obvyklejší mně (i z těch pravidel) přijde varianta s mezerou. Ještě je pro mě důležité, aby to bylo na celém webu stejně, což by opět nahrávalo použití toho standardního API s minimem ručních zásahů (které každý udělá trochu jinak) – viz třeba [detail události](https://cesko.digital/events/show-and-tell-8), kde datum a čas máme taky.

Já to každopádně mergnu a pokud nebudeš vysloveně proti, můžu to formátování poštelovat ex post.